.buildkite/cache-builder.yml

```diff
@@ -0,0 +1,35 @@
+# Nodes with values to reuse in the pipeline.
+common_params:
+  # Common plugin settings to use with the `plugins` key.
+  - &common_plugins
+    - automattic/bash-cache#v1.4.0
+    - automattic/git-s3-cache#v1.1.0:
+        bucket: "a8c-repo-mirrors"
+        repo: "wordpress-mobile/wordpress-ios/"
+  # Common environment values to use with the `env` key.
+  - &common_env
+    IMAGE_ID: xcode-12.5.1
+
+steps:
+
+  #################
+  # Build the CocoaPods Base Cache
+  #################
+  - label: ":cocoapods: Rebuild CocoaPods cache"
+    command: |
+      echo "--- :rubygems: Setting up Gems"
+      install_gems
+
+      echo "--- :cocoapods: Rebuilding Pod Cache"
+      cache_cocoapods
+    env: *common_env
+    plugins: *common_plugins
+
+  #################
+  # Build the CI cache
+  #################
+  - label: ":git: Rebuild git cache"
+    command: "cache_repo a8c-repo-mirrors"
+    plugins: *common_plugins
+    agents:
+      queue: "default"
```

Comment:
Can you explain to a Buildkite noob like me when this YML / pipeline would be triggered?

I don't see it referenced in the main pipeline, nor being run in https://buildkite.com/wordpress-mobile/wordpress-ios/builds/1977… so I'm wondering when this is used and how it works, and also why this is necessary.

I assumed that, as for most cases with CI and how we did it in CircleCI, the cache would instead be saved after install (and only if the lockfile differs), as typically we have steps like this in typical CI pipelines:
```
 - restore_cache key:{checksum Podfile.lock} # restore cache if one exists for that key, nothing if cache miss
 - pod install
 - save_cache key:{checksum Podfile.lock} # does nothing if cache for that key already exists, creates if it not
```

So I'm unsure why this is done differently, with the building of the cache handled as an entire separate pipeline that which don't see when it's triggered, rather than doing it the usual way and bundling that restore/build/save dance directly as a buundled custom command as part of the bash-cache plugin?

Reply:
For sure – it would be triggered via Buildkite's [Scheduled Jobs](https://buildkite.com/wordpress-mobile/wordpress-ios/settings/schedules) feature to run nightly (or on whatever frequency makes sense)

In terms of the triggering – the git cache has to be managed separately since it's based on a git mirror, which is a different kind of checkout – for that reason we can't just save the result of a regular build.

For Pods, I want to regenerate those from scratch from time to time in order to prevent the cache from containing every version we've ever used of the Pods. I think there's a middle ground here where we maintain some recent-ish versions, but the returns there are relatively small. Of course we could do this as part of the regular job, but the upload and packaging time for these caches is non-zero, so doing it async speeds up every job.

Reply:
Also as an aside, implementing this in this way was a good way to get set up for the "build via API call" – it'll use a similarly dedicated pipeline specifically for that task.

Reply:
> For sure – it would be triggered via Buildkite's Scheduled Jobs feature to run nightly (or on whatever frequency makes sense)

Ah! I thought about that, but then went to look at the Scheduled Jobs settings and didn't see it listed there so thought that I was mistaken and that wasn't where it was being used. But now I realize it's just not there _yet_ – because the pipeline needs to land in `develop` before you can add it there 😅 

> In terms of the triggering – the git cache has to be managed separately since it's based on a git mirror, which is a different kind of checkout – for that reason we can't just save the result of a regular build.

Oh I thought about that at first but from our 1:1 the other day I thought only Tumblr/Orangina used a git mirror while other, less "massive" repos would directly pull from GitHub without an intermediate mirror. It seems you decided to apply the mirror solution to every repo after all, because why not, right? 👍 

> For Pods, I want to regenerate those from scratch from time to time in order to prevent the cache from containing every version we've ever used of the Pods.

I see, and that's indeed a good idea. Could be worth maybe add some comments in that yml file to quickly mention/explain this whole rational?

Reply:
> apply the mirror solution to every repo

Probably just for the root projects, but not the Pods (WPiOS is > 2GB, so it's worth it, while the pods are all tiny).

But it's easy enough to add to any project that once it hits a point where it's a little slow we can add it to that project without much fuss!

Reply:
> Also as an aside, implementing this in this way was a good way to get set up for the "build via API call" – it'll use a similarly dedicated pipeline specifically for that task.

Yep I also thought that this pipeline would only be triggered by API call but then I saw no change in the `Fastfile` to provide a way to use the BuildKite API to do that like the ones we have to trigger CircleCI builds, so thought "maybe not, otherwise he'd have added a way to call that pipeline at least in order to test it".

---

Which I actually think could be nice before merging that PR (I mean, maybe not implement the whole BuildKite API call lane, because that logic will likely be implemented in the release-toolkit instead, but triggering that pipeline at least once before merging the PR just to validate that it works and passes too)

Reply:
> add some comments in that yml file

Ya good call – added in fbaefc6106673abbc3704fa5313f607dd8ad8c09